daemon/containerd/cache.go

```diff
@@ -2,211 +2,238 @@ package containerd
+	var layerDigest digest.Digest
+	if extraLayer != "" {
+		info, err := findContentByUncompressedDigest(ctx, c.is.client.ContentStore(), digest.Digest(extraLayer))
 		if err != nil {
-			return "", err
+			return "", fmt.Errorf("failed to find content for diff ID %q: %w", extraLayer, err)
 		}
-		lenHistory = len(parent.History)
+		layerDigest = info.Digest
 	}
-	for _, target := range ic.images {
-		if !isValidParent(target, parent) || !isValidConfig(cfg, target.History[lenHistory]) {
-			continue
-		}
-		return target.ID().String(), nil
+
+	var parentRef string
+	if parent != nil {
+		parentRef = parent.ID().String()
+	}
+	img, err := c.is.CreateImage(ctx, data, parentRef, layerDigest)
+	if err != nil {
+		return "", fmt.Errorf("failed to created cached image: %w", err)
 	}
 
-	return "", nil
+	return image.ID(img.ImageID()), nil
 }
 
-func isValidConfig(cfg *container.Config, h image.History) bool {
-	// todo: make this format better than join that loses data
-	return strings.Join(cfg.Cmd, " ") == h.CreatedBy
+func (c cacheAdaptor) IsBuiltLocally(target image.ID) (bool, error) {
+	ctx := context.TODO()
+	value, err := c.is.getImageLabelByDigest(ctx, target.Digest(), imageLabelClassicBuilderContainerConfig)
+	if err != nil {
+		return false, fmt.Errorf("failed to read container config label: %w", err)
+	}
+	return value != "", nil
 }
 
-func isValidParent(img, parent *image.Image) bool {
-	if len(img.History) == 0 {
-		return false
-	}
-	if parent == nil || len(parent.History) == 0 && len(parent.RootFS.DiffIDs) == 0 {
-		return true
-	}
-	if len(parent.History) >= len(img.History) {
-		return false
-	}
-	if len(parent.RootFS.DiffIDs) > len(img.RootFS.DiffIDs) {
-		return false
-	}
+func (c cacheAdaptor) Children(id image.ID) []image.ID {
+	ctx := context.TODO()
 
-	for i, h := range parent.History {
-		if !reflect.DeepEqual(h, img.History[i]) {
-			return false
-		}
-	}
-	for i, d := range parent.RootFS.DiffIDs {
-		if d != img.RootFS.DiffIDs[i] {
-			return false
+	if id.String() == "" {
+		imgs, err := c.is.getImagesWithLabel(ctx, imageLabelClassicBuilderFromScratch, "1")
+		if err != nil {
+			log.G(ctx).WithError(err).Error("failed to get from scratch images")
+			return nil
 		}
+		return imgs
 	}
-	return true
-}
 
-func (ic *imageCache) isParent(ctx context.Context, img *image.Image, parentID image.ID) bool {
-	ii, err := ic.imageService.resolveImage(ctx, img.ImageID())
+	imgs, err := c.is.Children(ctx, id)
 	if err != nil {
-		return false
-	}
-	parent, ok := ii.Labels[imageLabelClassicBuilderParent]
-	if ok {
-		return parent == parentID.String()
+		log.G(ctx).WithError(err).Error("failed to get image children")
+		return nil
 	}
 
-	p, err := ic.imageService.GetImage(ctx, parentID.String(), backend.GetImageOpts{})
-	if err != nil {
-		return false
+	return imgs
+}
+
+func findContentByUncompressedDigest(ctx context.Context, cs content.Manager, uncompressed digest.Digest) (content.Info, error) {
```

Comment:
Sucks that we need to walk the content store for this 😓

Reply:
Yeah, at least we have the label filter, so it's not as bad as just `Walk`, but still 😞

Reply:
Silly question; this is to find compressed content based by the content pre compressing? What happens if there's multiple matches? (thinking here that images with different compression could have the same uncompressed content). Could that result in random results?

Reply:
If the content store contains exactly the same layer but compressed with different compression algorithms (for example gzip and zstd) then yes, potentially the result would be random (it could either use gzip or zstd).

Considering this is a legacy builder, perhaps it would make sense to prefer gzip compressed blobs first?

Reply:
`gzip` is the most portable, as it's required to be supported, so should always work.

In general, the _mixed compression_ scenario is the most problematic (some layers `gzip`, other layers `zstd`) because in that situation a client that doesn't support `zstd` can ignore half of the layers.

Of course there could be a situation where a user _wants to_ use `zstd`, so that could still be an option, but if it does, then _all layers_ should be `zstd`.

So I guess the "ideal" would be to make all layers match the same compression (if the first one was `zstd`, then prefer that).

That might be tricky though if the first match is "random", so preferring `gzip` would _probably_ be a good start (??); in that case the first layer would most likely be `gzip`, and follow-up layers as well (because of ordering).

Perhaps the only "full" solution would be having to do 2 passes;

- collect all possible options (with their compression)
- if there's mixed compression in some of them, then.. pick one that's available for all of them
- as a last resort, allow mixed compression

That's probably much too complicated though, but I'm wondering what BuildKit does in this scenario, or does it always recompress layers if needed?

Reply:
Note that this only affects the `--cache-from` which already can result in picking up images that are not 100% matching.

Not sure how Buildkit handles it, I think `cache-from` behaves differently anyway, and even then  it has its own cache. With legacy builder we directly use images from the graph drivers and they're all stored in uncompressed form.

Considering that the legacy builder is deprecated anyway, I don't think it's worth to do anything more than just preferring to use the gzip.

Reply:
> Note that this only affects the --cache-from which already can result in picking up images that are not 100% matching.

Good point  as well, also `--cache-from` with the classic builder has always been considered a "best effort" (but potentially insecure / false matches), so ... probably fine?

> Ah, forgot that the `content.Info` doesn't provide information about the content media type 🤦🏻 So we can't really determine if it's a gzip or not without reading it first.

Bummer!

Ah, I guess that makes sense if it follows the same approach as registry "blobs"? (i.e., the blob itself is just "data", and in that case the manifest defines how the data should be interpreted), or does it store that information, but just doesn't expose it?

Reply:
No, it's not stored at all.
Although technically the content store writer interface allows to pass a descriptor: https://github.com/containerd/containerd/blob/90ec00c80cbdb4f2f031259a2415f519313fbef2/core/content/content.go#L188-L199

Also, `ReaderAt` accepts a descriptor, not just a digest, which hypothetically would mean that in future it _could_ take a media type into account and fail to return a reader if the media types doesn't match. But currently `MediaType` is ignored: https://github.com/containerd/containerd/blob/90ec00c80cbdb4f2f031259a2415f519313fbef2/core/content/proxy/content_store.go#L116-L129

Reply:
Yeah, not sure if it can, because how to interpret content is a N:1 relation with the content itself. For example, the manifest-index below is perfectly valid; The referenced images manifests can be interpreted as either an OCI image or a "Docker" image; it's the same content but can be interpreted with different semantics. Similar situations could happen for other content;

```
docker buildx imagetools inspect FOOBAR
Name:      docker.io/library/FOOBAR
MediaType: application/vnd.oci.image.index.v1+json
Digest:    sha256:77906da86b60585ce12215807090eb327e7386c8fafb5402369e421f44eff17e

Manifests:
  Name:      docker.io/library/FOOBAR:latest@sha256:aa772c98400ef833586d1d517d3e8de670f7e712bf581ce6053165081773259d
  MediaType: application/vnd.oci.image.manifest.v1+json
  Platform:  linux/amd64

  Name:      docker.io/library/FOOBAR:latest@sha256:7409efd2c351d36aaca162069e56a19fa2633944215cc478832a72d7eadfaf10
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/amd64
```